pkg/compute/kube/client.go

```diff
@@ -442,10 +464,117 @@ func (c *client) GetProxyStatsHeaders(p *models.Proxy) (map[string]string, error
+				pod.ObjectMeta.UID, pod.Namespace, svcAccount.Namespace)
+		return nil, errNamespaceDoesNotMatchProxy
+	}
+
+	// Ensure the Name encoded in the certificate matches that of the Pod
+	// TODO(draychev): check that the Kind matches too! [https://github.com/openservicemesh/osm/issues/3173]
+	if pod.Spec.ServiceAccountName != svcAccount.Name {
+		// Since we search for the pod in the namespace we obtain from the certificate -- these namespaces will always match.
+		log.Warn().Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrFetchingPodFromCert)).
+			Msgf("Pod with UID=%s belongs to ServiceAccount=%s. The pod's xDS certificate was issued for ServiceAccount=%s",
+				pod.ObjectMeta.UID, pod.Spec.ServiceAccountName, svcAccount)
+		return nil, errServiceAccountDoesNotMatchProxy
+	}
+
+	return &pod, nil
+}
+
+// getTelemetryPolicy returns the Telemetry policy for the given proxy instance.
+// It returns the most specific match if multiple matching policies exist, in the following
+// order of preference: 1. selector match, 2. namespace match, 3. global match
+func (c *client) getTelemetryPolicy(proxy *models.Proxy) *policyv1alpha1.Telemetry {
+	pod, _ := c.GetPodForProxy(proxy)
+	if pod == nil {
+		return nil
+	}
+
+	var policy *policyv1alpha1.Telemetry
+
+	for _, t := range c.kubeController.ListTelemetryPolicies() {
+		// If there is a global policy and a more specific policy hasn't been
+		// found yet, consider the global policy as a candidate
+		if policy == nil && t.Namespace == c.kubeController.GetOSMNamespace() {
+			policy = t
+			continue
+		}
+
+		// If the policy matches the namespace of the proxy's pod,
+		// consider this policy to be a candidate, but continue
+		// to look for a more specific policy that matches the pod
+		// based on a selector
+		if t.Namespace == pod.Namespace {
+			policy = t
+		}
+
+		// Look for a more specific match based on pod selector on the Telemetry resource.
+		// If we find a Telemetry resource that matches the pod's selector, this is
+		// the best match for this proxy.
+		selector := t.Spec.Selector
+		if len(selector) == 0 {
+			continue
+		}
+		sel := labels.Set(selector).AsSelector()
+		if sel.Matches(labels.Set(pod.Labels)) {
+			return t
+		}
+	}
+
+	return policy
+}
+
+// GetTelemetryConfig returns the Telemetry config for the given proxy instance.
+// It returns the most specific match if multiple matching policies exist, in the following
+// order of preference: 1. selector match, 2. namespace match, 3. global match
+func (c *client) GetTelemetryConfig(proxy *models.Proxy) models.TelemetryConfig {
+	var config models.TelemetryConfig
+
+	config.Policy = c.getTelemetryPolicy(proxy)
+
+	if config.Policy != nil && config.Policy.Spec.AccessLog != nil && config.Policy.Spec.AccessLog.OpenTelemetry != nil {
+		config.OpenTelemetryService = c.kubeController.GetExtensionService(config.Policy.Spec.AccessLog.OpenTelemetry.ExtensionService)
+	}
+
+	return config
```

Comment:
It's not clear why these functions are being moved. Could you describe this further?

Reply:
the k8s package is intended to not know anything about OSM objects. It's intent is to operate strictly on K8s objects, while the compute/kube package is aware of how to translate between the 2.

The k8s package should have simple crud-like methods for each object for Get, Update, List, Create, etc.

Reply:
GetTelemetryConfig calls getTelemetryPolicy, which calls getPodForProxy. Since GetPodForProxy is moving to the kube client, I think it makes sense for the GetTelemetry methods to move there also - otherwise we would need to call computeClient methods from the k8s client.

Reply:
I would argue these Telemetry related methods are unrelated to "Compute". These need to sit in the same pkg as other `Getters` (e.g. GetUpstreamTrafficSetting). If you look at this API, it does an aggregated GET on OSM API resources given some input. 

Also moving the getPodForProxy method should not be a reason to move the Telemetry helpers. Putting this into the "Compute" interface seems highly misleading and hard to reason.

Reply:
It's just about walls of abstraction. The plan is to not leak OSM resources into the k8s package, which is for simple crud operations. This allows us to hide complexities of informers, or non-informer based crud operations here, while the compute package is responsible for translating to/from OSM <-> k8s.

The entire purpose of that package is exactly the reason it should be placed there. Maybe compute is a bad name, and we can look to change it?

Note that the `GetUpstreamTrafficSetting` is exported in the Compute package as well.

Reply:
@steeling I get your point. MeshCatalog was meant to translate k8s objects to internal OSM objects. It seems like we have this new Compute interface doing the same. Maybe those can be collapsed into the Catalog?

Reply:
Ya I agree the 2 are a bit similar, but let me add some more reasoning and we can decide how we want to proceed from there.

Let's assume the following:
1. Interfaces exist to decouple from implementations, allowing us to leverage other implementations in the future.
2. With that in mind, it makes the most sense to push as much business logic out of specific implementations, to allow that business logic to apply to all instances of an interface. This prevents us from having to reimplement business logic.

With the above assumptions, the intent of each package is:

1. The k8s.Controller (package `k8s`): abstracts *how* we get objects from kubernetes. This may be an informer, a direct query to the k8s master, or, for example in a future where we leverage remote control planes, a local cache of k8s objects.
2. The compute client abstracts the *what*, that is, it abstracts Kubernetes itself. It allows a future where we may leverage running OSM outside of kubernetes
3. The mesh catalog contains just the business logic. With that, there should never be multiple instances, and we should remove the interface itself, and have everything directly rely on the MeshCatalog (as opposed to the MeshCatalogger).

This allows us to not need to re-implement the business logic that exists in the mesh catalog if we were to move off of kubernetes, or query kubernetes in a different manner.

Reply:
The mesh catalog was meant to abstract Kubernetes. So I suggest we simplify these layers of abstraction if the catalog and compute client are meant to do the same. Currently, it's hard to make sense of what goes into the compute client vs the catalog, and it seems a bit arbitrary.

Reply:
The meshcatalog contains *a lot* of business logic. That would make the undertaking significantly bigger if we ever wanted to implement a non-k8s client, and we had to implement all the functionality of the mesh catalog. It seems much simpler to keep the layers I described above and *only* require an implementation of the logic specific to the compute client. I disagree that it's arbitrary, in fact there's a simple set of rules to decide:

1. If a method deals with both OSM and K8s objects -> compute client
2. If a method *only* deals with k8s objects -> k8s client
3. If a method *only* deals with OSM objects -> mesh catalog.

I suspect a better approach may be to begin deconstructing some of the methods in the meshcatalog (to eventually remove it), such that the xDS builders are more aware of the underlying fields. For example:

```go
// GetOutboundMeshClusterConfigs returns the cluster configs for the outbound mesh traffic policy for the given downstream identity
func (mc *MeshCatalog) GetOutboundMeshClusterConfigs(downstreamIdentity identity.ServiceIdentity) []*trafficpolicy.MeshClusterConfig {
	var clusterConfigs []*trafficpolicy.MeshClusterConfig

	for _, meshSvc := range mc.ListOutboundServicesForIdentity(downstreamIdentity) {
		meshSvc := meshSvc // To prevent loop variable memory aliasing in for loop

		// ---
		// Create the cluster config for this upstream service
		clusterConfigForServicePort := &trafficpolicy.MeshClusterConfig{
			Name:                          meshSvc.EnvoyClusterName(),
			Service:                       meshSvc,
			EnableEnvoyActiveHealthChecks: mc.GetMeshConfig().Spec.FeatureFlags.EnableEnvoyActiveHealthChecks,
			UpstreamTrafficSetting:        mc.GetUpstreamTrafficSettingByService(&meshSvc),
		}
		clusterConfigs = append(clusterConfigs, clusterConfigForServicePort)
	}

	return clusterConfigs
}
```

The xDS builders/generators could really do the following:

```go
for _, svc := range c.ListOutboundServicesForIdentity(identity) {
   builder.AddOutboundService(svc, c.GetUpstreamTrafficSettingByService(&svc))
}
builder.SetEnableEnvoyHealthChecks(c.GetMeshConfig().Spec.FeatureFlags.EnableEnvoyActiveHealthChecks)
```

This removes the need to have an intermediate `MeshClusterConfig` type, as well as the method in the mesh catalog (and maybe eventually the mesh catalog itself).

I've called this out in #4926 and #4863, but I think time will tell more how much we can remove